src/item_location.cpp

```diff
@@ -344,37 +398,74 @@ item_location::item_location( const vehicle_cursor &vc, item *which )
 
 bool item_location::operator==( const item_location &rhs ) const
 {
-    return ( ptr ? ptr->what : nullptr ) == ( rhs.ptr ? rhs.ptr->what : nullptr );
+    return ( ptr ? ptr->target() : nullptr ) == ( rhs.ptr ? rhs.ptr->target() : nullptr );
 }
 
 bool item_location::operator!=( const item_location &rhs ) const
 {
-    return ( ptr ? ptr->what : nullptr ) != ( rhs.ptr ? rhs.ptr->what : nullptr );
+    return ( ptr ? ptr->target() : nullptr ) != ( rhs.ptr ? rhs.ptr->target() : nullptr );
 }
 
 item_location::operator bool() const
 {
-    return ptr && ptr->what;
+    return ptr && ptr->target();
 }
 
 item &item_location::operator*()
 {
-    return *ptr->what;
+    return *ptr->target();
 }
 
 const item &item_location::operator*() const
 {
-    return *ptr->what;
+    return *ptr->target();
 }
 
 item *item_location::operator->()
 {
-    return ptr->what;
+    return ptr->target();
 }
 
 const item *item_location::operator->() const
 {
-    return ptr->what;
+    return ptr->target();
+}
+
+void item_location::serialize( JsonOut &js ) const
+{
+    if( ptr ) {
+        ptr->serialize( js );
+    } else {
+        js.start_object();
+        js.member( "type", "null" );
+        js.end_object();
+    }
+}
+
+void item_location::deserialize( JsonIn &js )
+{
+    auto obj = js.get_object();
+    auto type = obj.get_string( "type" );
+
+    int idx = -1;
+    tripoint pos = tripoint_min;
+
+    obj.read( "idx", idx );
+    obj.read( "pos", pos );
+
+    if( type == "character" ) {
+        ptr.reset( new item_on_person( g->u, idx ) );
+
+    } else if( type == "map" ) {
+        ptr.reset( new item_on_map( pos, idx ) );
+
+    } else if( type == "vehicle" ) {
+        auto *veh = g->m.veh_at( pos );
+        auto cur = vehicle_cursor( *veh, obj.get_int( "part" ) );
```

Comment:
Could also verify whether the `part` is valid - that is, if it is less than `veh->parts.size()`.